crates/ty_ide/src/inlay_hints.rs

```diff
@@ -1,62 +1,139 @@
+            label,
+        }
+    }
+
+    pub fn display(&self) -> InlayHintDisplay<'_> {
+        InlayHintDisplay { inlay_hint: self }
     }
 }
 
-#[derive(Debug, Clone, Eq, PartialEq)]
-pub enum InlayHintContent<'db> {
-    Type(Type<'db>),
-    CallArgumentName(String),
+#[derive(Debug, Clone)]
+pub enum InlayHintKind {
+    Type,
+    CallArgumentName,
+}
+
+#[derive(Debug, Clone)]
+pub struct InlayHintLabel {
+    pub parts: Vec<InlayHintLabelPart>,
+}
+
+impl InlayHintLabel {
+    pub fn simple(s: impl Into<String>, target: Option<crate::NavigationTarget>) -> InlayHintLabel {
+        InlayHintLabel {
+            parts: vec![InlayHintLabelPart {
+                text: s.into(),
+                target,
+            }],
+        }
+    }
+
+    pub fn prepend_str(&mut self, s: &str) {
+        match &mut *self.parts {
+            [InlayHintLabelPart { text, target: None }, ..] => text.insert_str(0, s),
+            _ => self.parts.insert(
+                0,
+                InlayHintLabelPart {
+                    text: s.into(),
+                    target: None,
+                },
+            ),
+        }
+    }
+
+    pub fn append_str(&mut self, s: &str) {
+        match &mut *self.parts {
+            [.., InlayHintLabelPart { text, target: None }] => text.push_str(s),
+            _ => self.parts.push(InlayHintLabelPart {
+                text: s.into(),
+                target: None,
+            }),
+        }
+    }
 }
 
-impl<'db> InlayHintContent<'db> {
-    pub const fn display(&self, db: &'db dyn Db) -> DisplayInlayHint<'_, 'db> {
-        DisplayInlayHint { db, hint: self }
+impl From<String> for InlayHintLabel {
+    fn from(s: String) -> Self {
+        Self {
+            parts: vec![InlayHintLabelPart {
+                text: s,
+                target: None,
+            }],
+        }
     }
 }
 
-pub struct DisplayInlayHint<'a, 'db> {
-    db: &'db dyn Db,
-    hint: &'a InlayHintContent<'db>,
+pub struct InlayHintDisplay<'a> {
+    pub inlay_hint: &'a InlayHint,
 }
 
-impl fmt::Display for DisplayInlayHint<'_, '_> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        match self.hint {
-            InlayHintContent::Type(ty) => {
-                write!(f, ": {}", ty.display(self.db))
-            }
-            InlayHintContent::CallArgumentName(name) => {
-                write!(f, "{name}=")
-            }
+impl<'a> InlayHintDisplay<'a> {
+    pub fn new(label: &'a InlayHint) -> Self {
+        Self { inlay_hint: label }
+    }
+}
+
+impl fmt::Display for InlayHintDisplay<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> std::fmt::Result {
+        for part in &self.inlay_hint.label.parts {
+            write!(f, "{}", part.text)?;
         }
+        Ok(())
     }
 }
 
-pub fn inlay_hints<'db>(
-    db: &'db dyn Db,
+#[derive(Default, Debug, Clone)]
+pub struct InlayHintLabelPart {
+    pub text: String,
+
+    pub target: Option<crate::NavigationTarget>,
+}
```

Comment:
I would recommend to keep the field private and expose methods to provide the information.

Additionally, I think we should have a `From<&str>` / `From<String>` / `new(text: String)` implementation for `InlayHintLabelPart` and use builder pattern for the target like `with_target` because it's optional.